### DISCUSSION TITLE
Resolve original type object from a type string literal

### DISCUSSION THREAD

**Discussion by abelcheung** [python/typing#1382] [General]
Given a type as string, say `"int"`, I can get the original class easily since it's a builtin type and resolvable with `eval("int")`. But is there any way to get other types or generic aliases working, like `ForwardRef('Literal')._evaluate()` but can automatically determine where otherwise unresolvable names would come from?

I think it could be sort of attempted by importing common modules first (such as `collections.abc`, `typing` etc), or by exhaustively creating mapping of unresolvable common names and their corresponding modules. But I bet somebody might have created a better prior art, so it could be better to ask here first.

**Comment by hmc-cs-mdrissi** [comment]
You need knowledge of namespaces (globals/locals) where it should be evaluated. Usually for type annotations as string that come from class attribute/function signature you can try to fetch related namespace of that class/module. But if you have string with no context where it was used to define type you don't have enough information.

**Reply by abelcheung** [reply]
Hm, guessed this would be the case. Perhaps it's necessary to tell users to supply the context themselves then. Actually I'm dealing with inferred types from type checkers; `mypy` at least provided enough information like `typing.Literal[something]`, but `pyright` strips all the prefixes away when printing its output.
